### PR TITLE
bpo-25643: Refactor the C tokenizer

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue25643

Based on the original patch by Serhiy Storchaka

### CODE REVIEW

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c:234]:
```suggestion
        if (memcmp(cs, "utf-8", 5) != 0 && !set_readline(tok, cs)) {
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
        }
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
    }
    else {
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c:409]: Decref `line` before returning?

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
        char *line = Py_UniversalNewlineFgets(tok->inp,
                                              (int)(tok->end - tok->inp),
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
        if (tok->fp_interactive &&
            tok_concatenate_interactive_new_line(tok, line) == -1) {
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
static int
tok_concatenate_interactive_new_line(struct tok_state *tok, const char *line) {
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: `PyMem_REALLOC` is deprecated.

```suggestion
        newbuf = (char *)PyMem_Realloc(newbuf, newsize);
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: `PyMem_FREE` is deprecated.

```suggestion
            PyMem_Free(cs);
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: `PyMem_FREE` is deprecated.

```suggestion
            PyMem_Free(cs);
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: Shouldn't one keep a strong ref for `line` until done with `buf`? That is: decref after the call to `tok_concatenate_interactive_new_line` and before `return 1`?

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: For consistency:

```suggestion
        if (u == NULL) {
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: `PyMem_FREE` is deprecated.

```suggestion
            PyMem_Free(tok->buf);
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: `PyMem_FREE` is deprecated.

```suggestion
            PyMem_Free(newtok);
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c:876]: Need to free `newtok` after call to `tok_concatenate_interactive_new_line`?

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: Free with `NULL` pointer is a no-op.

```suggestion
        PyMem_Free(tok->buf);
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
    if (tok->start == NULL) {
        tok->cur = tok->inp = tok->buf;
    }
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
        if (!tok_readline_recode(tok)) {
            return 0;
        }
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
        if (!tok_readline_raw(tok)) {
            return 0;
        }
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
        if (tok->lineno > 2) {
            tok->decoding_state = STATE_NORMAL;
        }
        else if (!check_coding_spec(tok->cur, tok->end - tok->cur,
                                    tok, fp_setreadl))
        {
            return 0;
        }
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
    if (!tok->encoding
        && (tok->decoding_state != STATE_NORMAL || tok->lineno >= 2)) {
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
        }
        else if (tok->prompt != NULL) {
```

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]:
```suggestion
        }
        else {
```

**Review by erlend-aasland — COMMENTED** [review]
@pablogsal, @serhiy-storchaka: I added some PEP 7 comments, some comments about deprecated memory functions, and some decref / free control questions. Hope you don't mind.

File: Parser/tokenizer.c
Comment (pablogsal) [reply Parser/tokenizer.c:234]: Not convinced about this suggestion. If `cs` is `utf-8blablablac` this will compare equal when it shouldn't. Also, is less clear IMHO

Comment (erlend-aasland) [reply Parser/tokenizer.c:234]: Yes, you're right.

File: Parser/tokenizer.c
Comment (pablogsal) [reply Parser/tokenizer.c:876]: Maybe I am missing something but `newtok` is freed afterwards. The only case when is not freed is when is assigned to `tok->buf` and therefore is part of the buffer

File: Parser/tokenizer.c
Comment (pablogsal) [reply Parser/tokenizer.c]: Yep, I am going to reorganise a bit this part

File: Parser/tokenizer.c
Comment (pablogsal) [reply Parser/tokenizer.c:409]: Done by https://github.com/python/cpython/pull/25050/files#r602936998

Comment (pablogsal) [reply Parser/tokenizer.c:409]: Fixed in 71cdeedbae

File: Parser/tokenizer.c
Comment (pablogsal) [reply Parser/tokenizer.c]: Fixed in 71cdeed

**Comment by pablogsal** [comment]
> @pablogsal, @serhiy-storchaka: I added some PEP 7 comments, some comments about deprecated memory functions, and some decref / free control questions. Hope you don't mind.

Thanks a lot for your comments and for the review. I have applied most of the PEP7 and made some changes to error returning paths so they are a bit more clear.

File: Parser/tokenizer.c
Comment (erlend-aasland) [reply Parser/tokenizer.c:876]: Maybe I'm reading this wrong, but it's not freed before the return in line 877.

Comment (pablogsal) [reply Parser/tokenizer.c:876]: Ahhh, you mean in case that `tok_concatenate_interactive_new_line` fails! Yeah, that was missing. Thanks for clarifiying

Comment (erlend-aasland) [reply Parser/tokenizer.c:876]: Sorry, my bad. Yes, that's the case.

Comment (pablogsal) [reply Parser/tokenizer.c:876]: Fixed in 87626f3ab4

**Comment by pablogsal** [comment]
@erlend-aasland Once you are ok with the changes, you can approve the PR if you wish :)

File: Parser/tokenizer.c
Comment (erlend-aasland) [inline Parser/tokenizer.c]: One last nitpick :)

```suggestion
    else {
```

**Review by erlend-aasland — APPROVED** [review]
> @erlend-aasland Once you are ok with the changes, you can approve the PR if you wish :)

LGTM :)

**Pull request merged**
